Rewrite of the doc generation

ph opened the pull request:
**Motivation**: We have decided to rewrite the documentation generator to add more features
to our documentation and to fix some outstanding bugs.

This PR addresses the following problem:
- Hard to add new features to the code base.
- The code was tied with Logstash.
- No easy way to publish the doc for a single plugin
- The plugin author was not able to test their documentation changes
- The reported errors were hard to understand.
- No easy way to automate it.
- the minimal requirement on base classes.
# Features
## Generating the documentation for a Logstash instance

You can now generate the documentation for a Logstash distribution.

_prerequisite_:
- You need to have installed the plugins inside Logstash; this is necessary to pickup the right file when generating the documentation.
- You need to have the development gems installed. `bin/logstash-plugin --no-verify --development`

If you have all of the above you can run this command:

``` sh
rake docs:generate-plugins
rake doc:generate-plugins[/tmp/the-new-doc] # Depending on the shell you are using you might need to quote the task name, like this rake "doc:generate-plugins[/tmp/new-doc]"

```

This will give you an output similar to this

``` sh
logstash-input-file > SUCCESS
logstash-input-s3 > SUCCESS
logstash-input-kafka > FAIL
[...]
Exceptions: XXXX
```

The generator will try to generate the doc for all the plugins defined in the _Gemfile_ and installed in Logstash, if anything goes wrong it won't stop the generation of the other plugin. The Task will also report any errors with stack traces at the end, if one plugin fails the build,
you can interrupt the process, and it will output the current errors before exiting.

The docs will be generated in `build/docs/`.
## Generating the documentation for all the plugins from the organization

You can now generate the documentation from the master for all the plugin in the _logstash-plugins_ organization.

_prerequisite_
- You need to go in the `tools/logstash-docgen` directory
- You need to have the dependency installed.
  
  To get started you can run the following commands:
  
  ``` sh
  cd tools/logstash-docgen
  bundle install
  ```

You can use the the `bin/logstash-docgen` command to generate any plugin that you want; this executable can generate all the plugins or specific one from their master branch.

Usages:

``` sh
bin/logstash-docgen --all # will generate the doc for all the plugins
bin/logstash-docgen logtash-input-file logstash-input-s3 # generate doc for 2 plugins
```

**See:** `bin/logstash-docgen --help` for complete usage.

**Notes:**
- The nature of theses tasks require a lot of external execution to make sure all the doc are done in isolation, this process can take a long time.
- Some plugins will be skipped, see `logstash-docgen.yml` for details.
- The script will inject itself as a dependency on the plugin.

The doc are generate by default in a `target, where the command is executed.
## Testing the documentation as the plugin author

_prerequisite_
- Declare the `logstash-docgen` as development dependency
- Add `require "logstash/docgen/plugin_doc"` to the `Rakefile`
- Run `bundle install`

After you can have access to a few rake tasks, you can list them with `bundle exec rake -T`

```
bundle exec rake doc:asciidoc # return the raw asciidoc
bundle exec rake doc:html # Transform the asciidoc into html and open your default browser
bundle exec rake doc:html:raw # Give you the raw html
```

Fixes: #5762

ph commented:
@suyograo Ready for a run of review, the build are failling because `launchy` the gem I used to launch the browser is `ISC` licensed, I believe this is compatible with APL. If its not the case I can easily remove that feature.

I am working to add tests, but they arent easy, i will push another commit with them.

suyograo commented on rakelib/docs.rake:
```diff
@@ -1,25 +1,22 @@
-      cmd = "bin/bundle exec ruby docs/asciidoc_index.rb asciidoc_generated #{type}"
-      system(cmd)
-    end
+    FileUtils.mkdir_p(DOC_DIRECTORY)
```
Can we take a flag to configure the output directory?

suyograo commented on tools/logstash-docgen/README.md:
```diff
@@ -0,0 +1,80 @@
+**Features**
+
+## Generating the documentation for a Logstash instance
```
Nice! Thanks for putting up a README

suyograo commented:
@ph Couple of points as I test this:
* Can we have a single script in tools/docgen, for example -- `tools/docgen/from_installed_plugins.sh` which sets up all the prereqs? `rake bootstrap`, `rake test:install-core` (which install the dev dependencies). This script can be used in CI environment.
* I ran `rake docs:generate-plugins`. Checked beats input, the IDs for the config links are repeated. See `<<plugins-inputs-add_field-add_field>>` instead of `<<plugins-inputs-add_field>>`

suyograo commented on logstash-core/logstash-core.gemspec:
```diff
@@ -44,6 +44,10 @@ Gem::Specification.new do |gem|
 
+  # Required by the documentation tools
+  gem.add_runtime_dependency "asciidoctor", "~> 1.5.4"
+  gem.add_runtime_dependency "launchy", "~> 2.4.3"
```
Hmm, I am a bit iffy about `launchy`. If it's not essential, and the license is not in our list, i would vote to remove it. You can always launch this stuff from your browser manually, right?

ph replied:
Yup I will remove it, I think `rake doc:html | open chrome` might work?;)

suyograo commented on tools/logstash-docgen/lib/logstash/docgen/parser.rb:
wow :)

suyograo commented on tools/logstash-docgen/lib/logstash/docgen/parser.rb:
Thanks for all the comments. Really nice.

suyograo commented on tools/logstash-docgen/logstash-docgen.yml:
```diff
@@ -0,0 +1,8 @@
+---
+ignore_plugins:
```
Woohoo. nice

ph commented on tools/logstash-docgen/lib/logstash/docgen/parser.rb:
Well In an ideal world I would just pull the json and parse it... :(

But #5096 is really close to get merge.

ph commented:
@suyograo I have pushed the changes for your initial review. thanks!

ph commented:
@suyograo Also I put the **ci scripts** in the **ci folder**, I think it is better to have all of them together.

ph commented:
@suyograo I was missing a commit for the double id, the latest commit fixes the problem with the rabbitmq and also add the release date of the gem by querying the rubygems api.

ph commented:
Fetching the  compatible versions

```
=== collectd

Version: 3.0.2
Released on: 2016-07-14
https://github.com/logstash-plugins/logstash-codec-collectd/blob/master/CHANGELOG.md#302[Changelog]
Compatible: 5.0.0, 2.3.4, 2.3.3, 2.3.2, 2.3.1
```

ph commented:
Need another round of review.

Before merging we need to do the following:

- create the logstash-docgem and publish it.
- remove the **:path** in the Logstash's Gemfile (if we don't the artifact will be broken).

suyograo commented:
Will review today

suyograo commented on rakelib/test.rake:
```diff
@@ -39,6 +39,7 @@ namespace "test" do
 
   desc "run core specs"
   task "core" => ["setup"] do
+    puts $LOAD_PATH
```
Was this a left over debug statement?

suyograo commented on spec/fixtures/dummy.gemspec:
```diff
@@ -0,0 +1,24 @@
+  s.summary         = "This plugin is only used in the acceptance test"
+  s.description     = "This plugin is only used in the acceptance test"
+  s.authors         = ["Elasticsearch"]
+  s.email           = 'info@elasticsearch.com'
```
info@elastic.co and Elastic for author

suyograo commented on tools/logstash-docgen/lib/logstash/docgen.rb:
```diff
@@ -0,0 +1,7 @@
+require "logstash/docgen/version"
+
+module Logstash
```
LogStash instead of `Logstash`?

suyograo commented:
@ph the links are still wrong. 

They are now `plugins-inputs-add_field` instead of `plugins-inputs-beats-add_field`

ph commented:
@suyograo updated 

See the output.

https://gist.github.com/ph/017684cedac8401a35ed3fc906943304

Also I want this merged: https://github.com/elastic/logstash/pull/5096

suyograo commented:
Works like a charm. LGTM. Please squash before merging.

@ph Let's not wait for #5096, we can always update once that is merged.

wiibaa commented:
@ph another curious question: do you plan to make `logstash-docgen` as a `logstash-devutils` dependency or should it really be added in each plugin ?

ph commented:
@wiibaa It could be a runtime dependency of `logstash-devutils`, but we were thinking about removing devutils.

ph commented:
Actually if it's a dependency of devutils, it would actually make more sense.

ph commented:
@suyograo I cannot merged it before #5096,.because I have to release a version of this gem to make logstash use it as a dev dependency. Logstash currently use this gem in the gemfile with the `:path` that point to a local directory. When we create an artifact we dont include the `logstash-docgen` in the package this will make the package unable to run or install new plugins.


And without #5096 the gem will be broken since it will try to use a local path that doesn't exist.

jordansissel commented:
Ready to merge, @ph?

ph commented:
beats and jdbc are causing build error as reported by @suyograo https://gist.github.com/suyograo/24f05da8403262021c753ecfaef6b582

suyograo commented:
Also when I go to `tools/logstash-docgen` I get this error (even after running `bundle install`):

```
suyog@machine:~/ws/elastic/logstash/tools/logstash-docgen (pr/5720)$ bin/logstash-docgen logstash-input-beats
Processing, 1 plugins: beats
beats > FAIL
--------------------------------------------------------------------------------
FAILURE: beats
       	Exception: LogStash::Docgen::Plugin::CommandException message: cmd: bundler install --jobs 8 --quiet --path /tmp/vendor, stdout: Could not find gem 'flores (~> 0.0.6) java' in any of the gem sources listed in
your Gemfile or available on this machine.
, stderr:
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:104:in `run_in_directory'
       		/Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler.rb:237:in `with_clean_env'
       		/Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler.rb:447:in `with_env'
       		/Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler.rb:237:in `with_clean_env'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:99:in `run_in_directory'
       		org/jruby/RubyDir.java:341:in `chdir'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:98:in `run_in_directory'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:94:in `bundle_install'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:52:in `generate'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:175:in `generate'
       		org/jruby/RubyProc.java:281:in `call'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/task_runner.rb:27:in `run'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:174:in `generate'
       		org/jruby/RubyArray.java:1613:in `each'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:173:in `generate'
       		/Users/suyog/ws/elastic/logstash/tools/logstash-docgen/lib/logstash/docgen/runner.rb:27:in `execute'
       		/Users/suyog/.rvm/gems/jruby-1.7.25/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run'
       		/Users/suyog/.rvm/gems/jruby-1.7.25/gems/clamp-1.0.0/lib/clamp/command.rb:133:in `run'
       		bin/logstash-docgen:6:in `(root)'
```

suyograo commented:
with Kafka:

```
suyog@machine:~/ws/elastic/logstash/tools/logstash-docgen (pr/5720)$ bin/logstash-docgen logstash-input-kafka
Processing, 1 plugins: kafka
kafka > FAIL
--------------------------------------------------------------------------------
FAILURE: kafka
       	Exception: LogStash::Docgen::Plugin::CommandException message: cmd: bundle exec rake vendor, stdout: , stderr: Bundler::GemNotFound: Could not find rake-11.2.2 in any of the sources
      materialize at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/spec_set.rb:95
             map! at org/jruby/RubyArray.java:2446
      materialize at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/spec_set.rb:88
            specs at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/definition.rb:140
        specs_for at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/definition.rb:185
  requested_specs at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/definition.rb:174
  requested_specs at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/environment.rb:19
            setup at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/runtime.rb:14
            setup at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler.rb:95
           (root) at /Users/suyog/.rvm/gems/jruby-1.7.25/gems/bundler-1.12.5/lib/bundler/setup.rb:19
          require at org/jruby/RubyKernel.java:1040
           (root) at /Users/suyog/.rvm/rubies/jruby-1.7.25/lib/ruby/shared/rubygems/core_ext/kernel_require.rb:1
```

suyograo commented:
Ok, found another issue. For ES output -- configs are got from common_config.rb which is not getting reflected in the doc generator.

https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch/common_configs.rb#L141

https://www.elastic.co/guide/en/logstash/master/plugins-outputs-elasticsearch.html#plugins-outputs-elasticsearch-pipeline

ph commented:
@suyograo for the ES code I think the issue might be because of `mod.config`, the mod, I think it's not taken into consideration. The regexp expect `config` at the beginning.

ph commented:
I've fixed the issue with the elasticsearch output, when I was gathering source_location I was only taking into consideration module that defined instance methods, I was wrong.

I have also fixed a render issue when we use `type => :array`

suyograo commented:
LGTM

jordansissel commented:
Testing this --

```
⓿ localhost(~/projects/logstash/tools/logstash-docgen) pull/5720
% bundle install
Fetching gem metadata from https://rubygems.org/............
Fetching version metadata from https://rubygems.org/...
Fetching dependency metadata from https://rubygems.org/..
Resolving dependencies...
Bundler could not find compatible versions for gem "bundler":
  In Gemfile:
    bundler (~> 1.12)

  Current Bundler version:
    bundler (1.11.2)
This Gemfile requires a different version of Bundler.
Perhaps you need to update Bundler by running `gem install bundler`?

Could not find gem 'bundler (~> 1.12)' in any of the sources
```

ph commented:
@jordansissel 
I dont remember adding this dependency on a specific bundler version, It might be the generator shipped with `bundler` that automatically add a version requirement.
https://github.com/elastic/logstash/pull/5720/files#diff-b1c3263ad41e99f8f41a04d5b170d576R28

ph commented:
@jordansissel I just pushed a new commit that remove the constraints, can you give it another shot?

jordansissel commented:
docgen fails if my ruby is MRI. Possibly expected, but noting this just in case.

jordansissel commented:
New issue this time. I use JRuby 1.7.24 for my ruby, and do this:

```
# (in git root, toplevel of logstash repo)
% rake bootstrap
% cd tools/logstash-docgen
% bundle install
% bin/logstash-docgen logstash-input-stdin
Processing, 1 plugins: stdin
stdin > FAIL
--------------------------------------------------------------------------------
FAILURE: stdin
        Exception: LogStash::Docgen::Plugin::CommandException message: cmd: bundler install --jobs 8 --quiet --path /tmp/vendor, stdout: Could not find gem 'logstash-codec-plain java' in any of the gem sources listed
in your Gemfile or available on this machine.
, stderr:
                /home/jls/projects/logstash/tools/logstash-docgen/lib/logstash/docgen/github_generator.rb:104:in `run_in_directory'
                /home/jls/.rvm/gems/jruby-1.7.24/gems/bundler-1.11.2/lib/bundler.rb:219:in `with_clean_env'
                /home/jls/.rvm/gems/jruby-1.7.24/gems/bundler-1.11.2/lib/bundler.rb:199:in `with_original_env'
```

Logstash works, though `bin/logstash -e ""` reads from stdin happily.

ph commented:
@jordansissel Is you default ruby MRI or jruby? because the `logstash-docgen` tool will use a shellout and I am wondering if its executed in jruby or mri?

jordansissel commented:
@ph I've tried with both MRI and Jruby as default ruby :\

ph commented:
Few modifications:

- make sure we can generate the index
-make sure we can generate the doc locally?

ph commented:
See this Issue https://github.com/elastic/logstash/issues/6086 for the source of the description in the index page.

fbaligand commented:
Hi guys,
what is missing in this PR so that it can be merged ?

karmi commented:
Hi @ph, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/logstash/pull/5720.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

karmi commented:
Hi @ph, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/logstash/pull/5720.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

ph commented:
@suyograo I've rebased, squashed this PR and removed the logstash-docgen from the gemfile and created 2 ci scripts for the build system. Since this gem is not pushed on rubygems and we dont it ship in the package we have to make sure we don't declare it in the gemfile. The ci scripts will take of that at build time.

Building docs for the plugins installed installed in logstash works without any problems.
**BUT** building plugins from their master branch fails for some, I have to go through all of them and either fix the issue in the plugin itself or in the doc.

Can you give a last look using the `ci/ci_docs.sh` command and I will merge that in master and make other PRs for the other issues.

suyograo commented:
```
suyog@machine:~/ws/elastic/logstash (pr/5720)$ ci/ci_docs.sh
Invoking bundler install...
Plugin version conflict, aborting
rake aborted!
Bundler::VersionConflict: Bundler could not find compatible versions for gem "octokit":
  In Gemfile:
    logstash-docgen (>= 0) java depends on
      octokit (~> 4.3.0) java

    octokit (= 3.8.0) java
```

@ph this needs new constraints I guess

ph commented:
constrains updated.

ph commented:
@suyograo I've fixed the issue with the %PLUGIN% from the common plugin option, we now have a doc post process that will take care of theses values, the branch is also rebased with the current master.

suyograo commented:
@ph LGTM let's get this puppy merged
